omegaconf/base.py

```diff
@@ -85,6 +94,18 @@ def _get_node_flag(self, flag: str) -> Optional[bool]:
         return self._metadata.flags[flag] if flag in self._metadata.flags else None
 
     def _get_flag(self, flag: str) -> Optional[bool]:
+        cache = self.__dict__["_flags_cache"]
+        if cache is None:
+            cache = self.__dict__["_flags_cache"] = {}
+
+        ret = cache.get(flag, _MARKER_)
+        if ret == _MARKER_:
```

Comment:
I don't like relying on exceptions for core logic.
Since this is about caching, we are not expecting to miss a lot so it doesn't make a big difference (about 2% in the dict initialization benchmark which is not much).

A quick micro benchmark shows that using marker is 2x faster:

```python
def f1(cache, flag):
    try:
        cache[flag]
    except KeyError:
        pass

_MARKER_ = object()

def f2(cache, flag):
    ret = cache.get(flag, _MARKER_)
    if ret is _MARKER_:
        pass


def test_cache_try_except(benchmark: Any) -> None:
    benchmark(f1, {}, "foo")


def test_cache_with_marker(benchmark: Any) -> None:
    benchmark(f2, {}, "foo")
```

Results:
![image](https://user-images.githubusercontent.com/376455/105614847-ff93b780-5d80-11eb-96e1-f6fa656addb2.png)

Reply:
Interesting (though I doubt this kind of performance gap (100 ns) can matter here, especially since you're testing the rarer cache miss case. No problem if you prefer to keep `_MARKER_`.

Reply:
As I said, using an exception was about 2% slower in the real benchmark.
In any case, my preference to avoid exception is that it just feels wrong to use an exception for a non exceptional situation.

Reply:
It’s actually pretty common in Python. The accepted answer to this SO question gives a good overview I think: https://stackoverflow.com/questions/16138232/is-it-a-good-practice-to-use-try-except-else-in-python

That being said there are people who prefer to avoid it where possible and I respect that :) The usual way for dicts is more to check `if key in mydict` rather than use a special marker like you did, but I really don’t think it’s a big deal here.

Reply:
That would have been the implementation I chose if this was not all about performance.
In the normal path, the flag is going to be cached, which means for each flag access I will access the cache twice instead of once.

I did not add a benchmark testing access, but I am sure the flags cache is helping there as as well.

Reply:
Just to be sure, by "dict initialization benchmark " do you mean `test_omegaconf_create[large_dict]`?

Reply:
> Just to be sure, by "dict initialization benchmark " do you mean `test_omegaconf_create[large_dict]`?

I ran a few tests on this benchmark and couldn't see any significant difference (between your implementation vs try/except), but there is too much variance between runs to be sure (even with 100 rounds). Anyway, resolving this conversation as I have no strong opinion on this point.